File: ruby/swap_nodes_in_pairs.rb
# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val)
#     @val = val
#     @next = nil
#   end
# end

# @param {ListNode} head
# @return {ListNode}
def swap_pairs(head)
  return head if head.nil? or head.next.nil?
  odd, even = head, head.next

  # reverse first two nodes
  odd.next = even.next
  even.next = odd
  head = even

  while odd.next and odd.next.next
    tmp = odd
    odd = odd.next
    even = odd.next
    odd.next = even.next
    even.next = odd
    tmp.next = even
  end
  return head
end

File: python/swapNodesInPairs.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        pre = ListNode(None)
        pre.next = head
        tmp = pre
        while tmp.next and tmp.next.next:
            tmp.next = head.next
            head.next = head.next.next
            tmp.next.next = head
            tmp = tmp.next.next
            head = head.next
        return pre.next
